### ISSUE TITLE
Support multichannel audio

### ISSUE THREAD

**Issue by iver56** [asteroid-team/torch-audiomentations#2]
There are several ways of representing multichannel audio:

## Channels last

shape like [batch_size, num_samples, num_channels]

* Common when loading/saving audio from/to wav

## Samples last

shape like [batch_size, num_channels, samples]

* Convenient for convolution. E.g. for STFT, the time dimension is commonly the last dimension (in torchaudio, torch-stft and asteroid's STFT), according to @mpariente 
* "I always prefer channel first because in case one wants to apply an operation to each channel independently can do straight reshaping. Otherwise you have to transpose then reshape." - @popcornell
* Librosa uses this convention

---

I would choose samples last based on this information

**Comment by iver56** [comment]
How should we represent multichannel audio in spectrograms though? I would suggest channels last, since that is very common in computer vision, and is compatible with the way conv2d is implemented in pytorch and tensorflow

**Comment by mpariente** [comment]
For waveforms, I'd prefer `(batch, channels, samples)`

I would also suggest channel first for the multichannel spectrograms: 
```
audio : (batch, channels, samples) -> spec-like : (batch, channel, freqs, time)
```
This would at least be consistent with Asteroid's filterbank API and the vision of TF transforms as 1D convolutions. 
CMIIW but Conv2d is channel first by default in pytorch.

**Comment by iver56** [comment]
Thank you for your insights :) I haven't actually done much computer vision in pytorch, so I was probably wrong in my previous comment. Btw, Keras supports both channels_first and channels_last when running/training computer vision models. It's a config parameter one can set.

**Comment by mpariente** [comment]
According to [this docs page](https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html), this will also be the case for PyTorch. So it's a choice that we have to make.

**Comment by iver56** [comment]
One advantage of channels last is that it's easy to write a 2-channel spectrogram to PNG, since it's already in a compatible format 😄 

Example:

![darthuizen_audio__wavs_2662890367_382eaf83bd_0](https://user-images.githubusercontent.com/1470603/94010085-bcf7a980-fda5-11ea-9f94-48657a7d6e4f.png)

**Comment by iver56** [comment]
If we don't want to make this decision, I guess an "easy" way out would be to support both (have it as a parameter), and permute axes when needed.

**Comment by iver56** [comment]
Channels first pros:

* This is typically what you get out of conv1d-based STFT implementations in pytorch
* GPU (cuDNN) _without_ Nvidia tensor cores support (e.g. Nvidia Pascal and earlier): channels first is faster

Channels last pros:

* CPU with AVX/SSE: channels last is faster
* GPU (cuDNN) **with** Nvidia tensor cores support (e.g. new GPU generations like Nvidia Turing, Volta and Ampere): channels last is faster
* Same shape as typical image formats (e.g. in `Pillow`)

Some sources regarding performance:

* https://stackoverflow.com/questions/44774234/why-tensorflow-uses-channel-last-ordering-instead-of-row-major
* https://stackoverflow.com/questions/44280335/how-much-faster-is-nchw-compared-to-nhwc-in-tensorflow-cudnn
* https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html

Is there anything else that is worth mentioning here?

**Comment by popcornell** [comment]
Channel first pros: only need to reshape to batch dimension for applying an augmentation to each channel independently. 
Channel last you have to permute then reshape.

**Comment by iver56** [comment]
Another pro of channels first: It's consistent with waveforms having channels first

So I'm going to go ahead and call the shot (at least in the short term): Support channels first in spectrogram transforms.

Since multichannel audio is already a first-class citizen in torch-audiomentations, I'll close this issue now
